src/analyzer/Analyzer.cc

```diff
@@ -465,6 +465,17 @@ Analyzer* Analyzer::FindChild(const char* name) {
     return tag ? FindChild(tag) : nullptr;
 }
 
+void Analyzer::CleanupChildren() {
```

Comment:
This is great. Could you create a follow-up ticket to make a pass to have exiting implicit cleanup logic in `Analyzer` methods use this function instead? I feel the way the different states are queried feels error-prone to me, and I also always wonder about subtle differences in behavior between first cleaning up all children and then performing an operation, and cleaning up children while iterating and performing the operation (even if we rely on that we could have an internal function to clean up or perform the op to avoid the duplicated analyzer state checking).

Reply:
Not quite following, not sure I'd move the existing methods over to this because then we'd iterate twice. I agree that this is all error-prone of course, but that's how the code currently is.

Reply:
I meant adding a (`protected`?) method like this to `Analyzer` (with much better name):

```cpp
template<typename Fn>
void OpOrCleanup(Fn f) {
    AppendNewChildren();

    for ( auto i = children.begin(); i != children.end(); ) {
        if ( ! ((*i)->finished || (*i)->removing) ) {
            f(*i);
            ++i;
        }
        else
            i = DeleteChild(i);
    }

    // NOTE: `CleanupChildren` only calls this at the start, but most functions
    // doing implicit cleanup call it twice.
    AppendNewChildren();
}
```

A rewritten `ForwardStream` would then look like this:

```cpp
void Analyzer::ForwardStream(int len, const u_char* data, bool is_orig) {
    if ( output_handler )
        output_handler->DeliverStream(len, data, is_orig);

    OpOrCleanup([&](Analyzer* cur) { cur->NextStream(len, data, is_orig); });
}
```

Reply:
I see. But not convinced a ticket makes sense here. There are tons of places in Zeek where we could clean things up but we don't have a habit of creating tickets for them; would feel kind of random to me to start this here (my 2c; may be a topic for a broader discussion if we want such tickets).